### PR TITLE
Runs the docker container as user 1001 (unpriveleged) and group 0 (root)

### PR DESCRIPTION
**Pull request by stephenmoloney**
### What does this commit/MR/PR do?

- Runs the docker container as user 1001 (unpriveleged) and group 0
(root)

###  Why is this commit/MR/PR needed?

- Not to run as root
- Not to run as nginx but an arbitrary user id
- Facilitate kubernetes security context

###  TODO

- [x] Implement similar logic for debian images

### CODE REVIEW

**Comment by alessfg** [comment]
Thanks for the PR @stephenmoloney! I'll merge as soon as you implement similar changes to the Debian images 😄

**Comment by KevinWMatthews** [comment]
Hi @stephenmoloney ,

I still need to actually run it, but it looks pretty good.

Is it worth setting the User ID using a variable instead of hard-coding it? I think we could allow users to override the UID from the command line when they build the image.

**Comment by stephenmoloney** [comment]
@KevinWMatthews I like that idea. I'll look into implementing that.

Is anyone good with debian, not sure what changes should be made to match this on debian.

@alessfg, yes hold off on merging this until everything is ready. thanks.

**Comment by stephenmoloney** [comment]
@KevinWMatthews 

```
Is it worth setting the User ID using a variable instead of hard-coding it? I think we could allow users to override the UID from the command line when they build the image.
```

Only two ways come to mind for achieving this:

1. Make uid and ENV variable - but this will not work I think, I haven't checked but I don't think USER_ID can be an environment variable. ie

```
USER_ID  ${SOME_UID}
```
Building it in as a build argument is not great because people then have to build their own images.

2. Using gosu.

I think by using gosu, it is possible to achieve this with a ENV variable. But this means adding more stuff to the image, is everyone ok with that... it's worth it?

**Comment by alessfg** [comment]
> 1. Make uid and ENV variable - _**but this will not work**_, I haven't checked but I don't think USER_ID cannot be an environment variable. ie
> 
> ```
> USER_ID  ${SOME_UID}
> ```

We could make this work using the `ARG` directive, as explained here https://vsupalov.com/docker-build-time-env-values/

> 2. Using gosu.
> 
> I think by using gosu, it is possible to achieve this with a ENV variable. But this means adding more stuff to the image, is everyone ok with that... it's worth it?

I don't think using gosu is a good idea. It'd add another dependency to the image.

**Comment by KevinWMatthews** [comment]
Sorry, I keep commenting on the wrong PR... doh. Duplicating some of what alessfg said - he beat me to it.

We can use Docker's ARG instruction to do something like:

```
ARG username=nginx
ARG userid=1001
```

at the top of the Dockerfile, and then reference it elsewhere using ${userid}.

For example,

```
	&& adduser -D -S -h /var/cache/nginx -s /sbin/nologin --uid ${userid} ${username} \

# ...

USER ${userid}
```

If left untouched, the image be built with nginx and 1001, but the user can override this using something like:
```
$ docker build --build-arg username=my_user --build-arg userid=1042 -t <tag_name> .
```
It seems to work on my box, but it would be good for more than one person to try it.

I could put these changes on top of your pull request if that's easier for you, but I can't seem to figure out how....

**Comment by KevinWMatthews** [comment]
@alessfg , do we need to change the `CONFIG` variable to use:
```
		--user=${username} \
		--group=root \
```
They are both currently set to `nginx`.

`CONFIG` contains options that are passed to the configure script that is used to build nginx.

**Comment by stephenmoloney** [comment]
That requires the user to build their own image and not use `nginxinc/docker-nginx-unprivileged` image.

**Comment by alessfg** [comment]
I agree with you @KevinWMatthews. However, at the end of the day, I think we could also simply set the userid to 1001, username to nginx, and group to root and have users tweak the source code if they want different settings. The original goal of this image was to tweak the official NGINX Docker image to allow NGINX to be run from/by an unprivileged user. The less tweaks we make to accomplish that, the better.

**Comment by KevinWMatthews** [comment]
@alessfg , makes sense. You do have a better view of the scope of NGINX Dockerfiles in general, so I'm happy to defer to your judgement.

If you need help propagating changes or doing basic changes, I'm glad to assist.

**Comment by alessfg** [comment]
@KevinWMatthews @stephenmoloney I have made a few additional changes to the Alpine Dockerfiles and tweaked the Debian Dockerfiles to match what we've been discussing. Any feedback/thoughts?

**Comment by stephenmoloney** [comment]
I can't work on this for about 2 weeks.

@alessfg Thanks for the changes to the debian images.

I see your removed ownership of `/var/log/nginx/` `usr/share/nginx/html/` by uid 1001, I do wonder if that is required though (I'm not 100% sure without testing).

I think some way of testing this image in travis would be nice to verify things are working as expected.

Any ideas ?

**Comment by alessfg** [comment]
I agree that Travis would be nice - I'll see if I can get it deployed in the near future. For now I've been doing manual testing and all images seem to be working when run normally via Docker. I'll try to test them on OpenShift and merge this PR if I don't find any further bugs.

File: mainline/stretch/Dockerfile [hunk 1]
```diff
@@ -89,7 +89,8 @@ RUN set -x \
 	fi
 
 # implement changes required to run NGINX as an unprivileged user
-RUN chown -R :0 /var/cache/nginx \
+RUN usermod -u 1001 nginx \
+	chown -R 1001:0 /var/cache/nginx \
```
Comment (KevinWMatthews) [inline mainline/stretch/Dockerfile]: Missing the leading `&&`?
```suggestion
	&& chown -R 1001:0 /var/cache/nginx \
```

File: stable/stretch-perl/Dockerfile [hunk 2]
```diff
@@ -90,7 +90,8 @@ RUN set -x \
 	fi
 
 # implement changes required to run NGINX as an unprivileged user
-RUN chown -R :0 /var/cache/nginx \
+RUN usermod -u 1001 nginx \
+	chown -R 1001:0 /var/cache/nginx \
```
Comment (KevinWMatthews) [inline stable/stretch-perl/Dockerfile]: Also missing the leading `&&`?

```suggestion
	&& chown -R 1001:0 /var/cache/nginx \
```

File: stable/stretch/Dockerfile [hunk 3]
```diff
@@ -89,7 +89,8 @@ RUN set -x \
 	fi
 
 # implement changes required to run NGINX as an unprivileged user
-RUN chown -R :0 /var/cache/nginx \
+RUN usermod -u 1001 nginx \
+	chown -R 1001:0 /var/cache/nginx \
```
Comment (KevinWMatthews) [inline stable/stretch/Dockerfile]: One more.
```suggestion
	&& chown -R 1001:0 /var/cache/nginx \
```

File: mainline/stretch/Dockerfile (hunk 1 above)
Comment (KevinWMatthews) [reply mainline/stretch/Dockerfile]: Well, this is a new GitHub feature?

**Comment by KevinWMatthews** [comment]
This looks good after a cursory examination and a few simple test cases. Running as non-root, serves content, logs pip to stdout and stderr.

Added comments about fixing a typo - I think we're missing the leading `&&` in a few places. 

Is that a new feature on GitHub? I'm curious what you think about it. Seems a bit verbose....

Thanks!

**Comment by KevinWMatthews** [comment]
@alessfg I see. If you have to make your own commit, that's fine by me.

Thanks!

**Comment by alessfg** [comment]
Done. Going to merge and update the Dockerhub images asap.

**Comment by stephenmoloney** [comment]
:+1:

**Comment by KevinWMatthews** [comment]
Awesome. Thanks!

**Pull request merged**